platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/transform/package-info.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This package contains types for capturing data about the chain of variant transformations that
+ * are available to satisfy an artifact selection request.
+ * <p>
+ * These types are a more lightweight representation of the data in {@link org.gradle.api.internal.artifacts.transform.TransformedVariant TransformedVariant},
+ * that will not pose problems to serialization in {@link org.gradle.internal.operations.BuildOperation BuildOperation}s and traces.  They
```

Comment:
I think its odd to talk about serialization in the docs for all these types. Its true that serialization is the reason we're doing this work _now_, but I think the stronger reason for using these types is to avoid exposing internals (with all their state and such) in these failure types.

Its just that once we serialized these types, we realized that we were exposing internal stateful types. Serialization is the thing that made us realize our mistake, but IMO isn't the core reason we want these types. Even if serialization worked for the internal types, we'd _still_ want these lightweight wrappers. 

I think it would be better to mention that we don't want to expose internals and that we only want to expose the information necessary to diagnose failures to compute artifact transforms. The serialization failures are just a side effect of exposing internal types

Reply:
That's a good point, keeping the internal types internal is more important and should go first.  I want to keep the bit about serialization at the end though, as it's a useful breadcrumb and it seems easily to accidentally start complexifying these types and run into problems with this.  It is an important property of these types that is tested, but only indirectly.